### DISCUSSION TITLE
prototype의 장점이 무엇이길래 javascript에서 채택했을까요? 그렇다면 prototype은 무엇인가요?

### DISCUSSION THREAD

**Discussion by JUDONGHYEOK** [woowacourse-study/2022-woowahan-bansanghwe#61] [JavaScript]
# prototype

 흔히들 프로토타입을 자바스크립트의 기반이 되는 개념이라고 하죠! 하지만 대부분의 책들에서는 그저 프로토타입을 자바스크립트에서 상속을 위해 구현된 객체이다 정도로밖에 설명을 하는데요. 상속만을 위해 구현된 객체라면 왜 프로토타입이 자바스크립트의 기반이라고 표현을 하는 것일까요? 그렇다면 클래스 기반의 언어와는 무엇이 다르길래 자바스크립트는 프로토타입을 채용했고 그 장점은 무엇이 있을까요?

 해당 글은 제 뇌피셜 절반 여러 블로그들 절반으로 이루어진 내용이므로 필터를 통해 읽어 주시길 바라겠습니다!

## 프로토타입이란?
 본격적으로 자바스크립트의 이야기를 하기 전 프로토타입의 뜻부터 살펴보고 가시죠 프로토타입, 영어로 prototype은 원래 형태 또는 전형적인 예, 기초 또는 표준이라고 합니다. 프로토타입은 프로그래밍에서만 쓰이는 개념은 아니구요. 보통 시제품이 나오기 전의 제품의 원형을 표현하는데 사용한다고 합니다. 흔히들 디자인 쪽에서도 자주 사용되는 언어이죠. 그렇다면 이 프로토타입이라는 단어의 기원은 어떻게 되는걸까요? 

http://superjang.com/archives/1964

 프로토타입은 16세기 프랑스에서 "a primitive form, original, or model after which anything is formed," 즉 어떤 것이 만들어진 원형이라는 뜻을 갖고 "original, primitive"라는 뜻을 가진 그리스어 prōtotypon에서 유래되었는데요. 이 프로토타입을 이해하기 위해선 classfication을 알아야합니다. 프로토타입이란 개념이 classification을 반박하여 나온 이론이기 때문인데요. calssfication은 흔히들 개체의 속성을 기반으로 그룹을 같은 범주에 속하게 합니다. 이 그룹이라는 것은 결국엔 다른 클래스 기반 언어에서 클래스가 되는 것이고 속성은 프로퍼티가 되는 것이죠. 

 그렇다면 이에 반박되는 이론은 무엇일까요? 바로 사용이론입니다. 비트겐슈타인이 19세기에 내놓은 이론인데요. 본질이라고 부를만한 공통 성질은 존재하지 않고 가족 간의 유사성 즉 가족유사성만이 존재한다고 합니다. 우리 가족이 동생은 엄마의 눈과 코, 아빠의 입과 걸음걸이가 닮고 형이 아빠의 눈과 코, 엄마의 입과 걸음걸이가 닮았다고 가정한다면 가족은 공통 성질은 없지만 교차해서 닮은 부분은 존재하고 이때문에 우리는 가족이라고 가정을 하는 것이라고 하죠.

 때문에 무엇이든 그 의미는 자체로 규정되는 것이 아니라 인간의 사용에 따라 달라진다고 주장하죠 이 사용이론은 1970년경 프로토타입이론으로 넘어오게 되는데요. 1970년대 초 Eleanor Rosch가 개발한 과학이론인데요 이 이론은 사람들을 해당 범주의 프로토타입 또는 이상적인 표현을 기반으로 분류합니다. 예를 들어 개의 개념이 종종 모피, 꼬리 및 발로 특정지어집니다. 이러한 개의 특징이 늑대나 코요테도 충족시킬 수 있지만 이 동물들은 개의 프로토타입이 되지 않습니다.

 기본적인 프로토타입 이론은 인간이 가진 등급의 개념구조를 가진다라고 주장합니다. 인간이 무언가를 분류할 때 자연스럽게 높은 유사성은 토대로 등급을 매기는 것이죠. 즉 객체는 정의에 따라 구분되는 것이 아닌 보기로부터 범주화 되는 것이죠. 따라서 맞다 아니다의 이분법적인 개념보다는 몇가지 특징만 원형과 비교하여 특징이 다를 수록 먼 범주를 정의하는 것입니다.

이러한 사용이론과 프로토타입 이론은 처음부터 속성을 기반으로 그룹을 맞다 아니다로 정의하는 것이 아닌 어디서 어떻게 사용하느냐에 따라 범주가 달라지고 현실에 존재하는 것 중 가장 좋은 본보기를 원형으로 한다는 점입니다.

## 자바스크립트에서의 프로토타입이란?
그렇다면 이제 자바스크립트에서의 프로토타입의 정의를 다시 한 번 살펴볼까요? 다음은 제가 존경하는 정재남님의 코어자바스크립트에 정의된 프로토타입의 개념입니다.

> 프로토타입 기반언어에서는 어떤 객체를 원형(prototype)으로 삼고 이를 복제함으로써 상속과 비슷한 효과를 얻는다.

 프로토타입은 객체를 원형으로 삼고 즉 현실에서 가장 좋은 본보기를 원형으로 삼고 이를 복제하는 것입니다. 즉 자바스크립트에서 프로토타입으로 연결된 하위 객체의 프로퍼티는 상속된 프로퍼티를 변경해도 원형에는 영향을 받지 않습니다. 또한 자바스크립트에서는 프로토타입을 인스턴스로 지정이 가능하죠. 

``` javascript
function Dog(){
	this.leg = 4
    this.run = fuction(){
    	console.log("달리기")
    }
}

const 뽀삐 = new Dog();

function Cat(){
	this.bark = function(){ 
    	console.log("야옹")
  	}
}

const 야옹이 = new Cat();

야옹이.prototype = 뽀삐

야옹이.leg = 2;
```

 이렇게 프로토타입은 인스턴스를 기준으로 자유로운 상속이 가능합니다. 즉 개별 인스턴스 수준에서도 메소드와 변수를 자유롭게 추가할 수 있는 것이죠. 인스턴스가 변경되어도 그 원형은 변하지 않습니다. 즉 인스턴스에게 위임을 하는 것이죠. 인스턴스가 해당 행동을 반응 하지 못할 경우에 다른 객체로 메세지를 전달하는 것입니다.

 이렇게 살펴보면 프로토타입이론과 연결되는 모습을 확인할 수 있습니다. 프로토타입 기반 언어는 클래스와 객체를 구별하지 않고 단순히 객체를 기반으로 상속이 가능합니다. 모든 객체를 다른 객체의 프로토 타입으로 연결할 수 있습니다. 이는 Lexical Scope 개념과도 연결이 됩니다.

Lexical Scope 즉 어휘적인 스코프 그 함수가 어디서 정의되었는지에 따라 스코프가 결정된다는 뜻인데요. 즉 클래스를 정의할 필요없이 프로토타입기반언어는 런타임에서도 새로운 종류의 객체를 만들고 체이닝이 가능합니다. 따라서 자유롭게 상속이 이루어지고 변경될 수 있다는 뜻이죠

이처럼 프로토타입 상속 형태의 객체는 클래스를 정의해야 인스턴스를 생성할 수 있는 클래스 기반 언어와는 달리 자유롭게 객체 생성 및 체이닝이 가능합니다. 자바스크립트가 프로토타입 기반 언어라는 것은 이러한 프로토타입에서 모든 자바스크립트의 중요 개념들이 나오는 것이죠. Lexical Scope도 그 함수의 사용에 따라 스코프가 결정되는 사용 이론과 이어지게 되고 자바스크립트에서 중요한 클로저나 this와 같은 경우도 이와 이어지게 됩니다.

## 프로토타입의 채택 이유?

그렇다면 자바스크립트는 왜 프로토타입을 채택했을까요? 제 생각에는 단순하게 생각하자면 자바스크립트는 결국엔 브라우저에서 동작하기 위해 1996년 넷스케이프 웹브라우저인 Netscape Navigator 2에 탑재되기 위해 만들어진 언어입니다. 97년 ECMA에서 ES1이 등장하게 되고 수많은 개념들이 자바스크립트로 흡수 되었지만 프로토타입 기반 언어라는 사실은 변함이 없습니다. 브라우저 즉 동적인 HTML변경이나 스타일 변경을 위해 만들어진 언어입니다. 따라서 사용자 인터랙션이 무엇보다 많이 발생할 수 밖에 없는 언어입니다. 사용자 인터랙션은 자바스크립트의 런타임 환경과 연결이 되죠. 이러한 런타임 환경에서 자바스크립트 프로토타입의 높은 자유도는 효과적으로 작동할 수 밖에 없습니다. 만약에 자바스크립트에서 클래스기반 언어를 채택했다면 동적으로 객체를 생성하고 인터랙션에 대처하는데도 수많은 복잡한 로직이 추가될 수 밖에 없는 것이죠. 자바스크립트는 이를 위해 높은 자유도를 가진 프로토타입 기반 언어를 채택하였고 수많은 사용자 인터랙션에 대처할 수 있게 되었다고 생각합니다.

참고자료

https://medium.com/@limsungmook/%EC%9E%90%EB%B0%94%EC%8A%A4%ED%81%AC%EB%A6%BD%ED%8A%B8%EB%8A%94-%EC%99%9C-%ED%94%84%EB%A1%9C%ED%86%A0%ED%83%80%EC%9E%85%EC%9D%84-%EC%84%A0%ED%83%9D%ED%96%88%EC%9D%84%EA%B9%8C-997f985adb42

자바스크립트는 왜 프로토타입을 선택했을까

프로토타입으로 검색하면 으레 나오는 서두처럼 저 또한 자바스크립트를 처음 접했을 때 가장 당황스러웠던 게 프로토타입이었습니다.

medium.com
https://poiemaweb.com/js-introduction

Introduction | PoiemaWeb

자바스크립트는 HTML, CSS와 함께 웹을 구성하는 요소중 하나로 웹브라우저에서 동작하는 유일한 언어로 1995년 Brendan Eich(Nescape)가 Navigator 2를 위하여 개발한 웹페이지에 포함되는 스크립트 언어

poiemaweb.com
https://developer.mozilla.org/ko/docs/Learn/JavaScript/Objects/Object_prototypes

Object prototypes - Web 개발 학습하기 | MDN

Javascript에서는 객체를 상속하기 위하여 프로토타입이라는 방식을 사용합니다. 본 문서에서는 프로토타입 체인이 동작하는 방식을 설명하고 이미 존재하는 생성자에 메소드를 추가하기 위해 프

developer.mozilla.org

**Comment by greenblues1190** [comment]
ES6에서 클래스를 도입한 이유는 무엇일까요?

**Comment by byhhh2** [comment]
- 만약 고양이와 개 둘다에서 속성을 찾지 못하면 어쩌나?

- 프로토타입 언어에 클래스를 도입한 이유는?

- 하위 객체의 프로토타입 속성을 변경하면 상위 객체도 변경되지 않나? 그에 수반되는 단점은?

- 클로저도 사용되는 처에서 스코프가 결정된다고 할 수 있나?

- 클로저란?

- 프로토타입 이론은 왜 나오게 됐나요?

- 왜 private, public을 지원하지 않는지?

**Comment by byhhh2** [comment]
만약 고양이와 개 둘다에서 속성을 찾지 못하면 어쩌나?

**Comment by byhhh2** [comment]
프로토타입 언어에 클래스를 도입한 이유는?

**Comment by byhhh2** [comment]
하위 객체의 프로토타입 속성을 변경하면 상위 객체도 변경되지 않나? 그에 수반되는 단점은?

**Comment by byhhh2** [comment]
클로저도 사용되는 처에서 스코프가 결정된다고 할 수 있나?

**Comment by byhhh2** [comment]
클로저란?

**Comment by byhhh2** [comment]
프로토타입 이론은 왜 나오게 됐나요?

**Comment by byhhh2** [comment]
왜 private, public을 지원하지 않는지?
